File: java/src/main/java/problems1_100/StringToIntegerAtoi.java
package problems1_100;

import java.math.BigInteger;

enum StringToIntegerAtoi {
    DEFAULT {
        @Override
        int myAtoi(String s) {
            if (s.isEmpty()) {
                return 0;
            }
            s = s.trim();
            boolean negate = false;
            if (s.startsWith("-")) {
                s = s.substring(1);
                negate = true;
            } else if (s.startsWith("+")) {
                s = s.substring(1);
            }

            StringBuilder builder = new StringBuilder();
            for (char c : s.toCharArray()) {
                if (Character.isDigit(c)) {
                    builder.append(c);
                    continue;
                }
                break;
            }
            if (negate) {
                builder.insert(0, '-');
            }
            if (builder.length() == 0
                || (builder.length() == 1 && (builder.charAt(0) == '-' || builder.charAt(0) == '+'))
            ) {
                return 0;
            }

            BigInteger result = new BigInteger(builder.toString());
            if (
                result.compareTo(new BigInteger(String.valueOf(Integer.MIN_VALUE))) < 0
            ) {
                return Integer.MIN_VALUE;
            } else if (
                result.compareTo(new BigInteger(String.valueOf(Integer.MAX_VALUE))) > 0
            ) {
                return Integer.MAX_VALUE;
            }
            return result.intValue();
        }
    };

    abstract int myAtoi(String s);
}

File: kotlin/src/main/kotlin/problems1_100/StringToIntegerAtoi.kt
package problems1_100

enum class StringToIntegerAtoi {
    DEFAULT {
        override fun myAtoi(s: String): Int {
            var s = s

            if (s.isEmpty()) {
                return 0
            }
            s = s.trim { it <= ' ' }
            var negate = false
            if (s.startsWith("-")) {
                s = s.substring(1)
                negate = true
            } else if (s.startsWith("+")) {
                s = s.substring(1)
            }

            s =
                buildString {
                    for (c in s.toCharArray()) {
                        if (c.isDigit()) {
                            append(c)
                            continue
                        }
                        break
                    }
                    if (negate) {
                        insert(0, '-')
                    }
                    if (isEmpty() ||
                        (length == 1 && (this.first().let { it == '-' || it == '+' }))
                    ) {
                        return 0
                    }
                }

            val result = s.toBigInteger()
            if (result < Int.MIN_VALUE.toBigInteger()) {
                return Int.MIN_VALUE
            } else if (result > Int.MAX_VALUE.toBigInteger()) {
                return Int.MAX_VALUE
            }
            return result.toInt()
        }
    },
    ;

    abstract fun myAtoi(s: String): Int
}
